File: modulo07-ferramentas-de-ia-para-gestao-de-projetos/modulo-04-estimativas-e-previsoes/monte-carlo-routewise.py
"""
Monte Carlo — RouteWise MVP
Módulo 4.2: Estimativas e Previsões Probabilísticas
"""

import random
import math


def triangular(o, m, p):
    """Sorteia um valor da distribuição triangular com parâmetros O, M, P."""
    u = random.random()
    fc = (m - o) / (p - o)
    if u < fc:
        return o + math.sqrt(u * (p - o) * (m - o))
    else:
        return p - math.sqrt((1 - u) * (p - o) * (p - m))


# Estimativas O / M / P em semanas
historias = {
    "US-01 Alertas de Velocidade":      (2, 3, 5),
    "US-03 Score de Comportamento":     (3, 4, 6),
    "US-04 Sensor de Abertura de Bau":  (1, 2, 3),
    "US-09 Carga Refrigerada":          (2, 3, 5),
}

HARDWARE_SEMANA = 9   # US-04 e US-09 so comecam quando o hardware chega
N = 10_000            # numero de simulacoes


totais = []
for _ in range(N):
    # Trilha software (comeca na semana 1)
    # US-03 depende de US-01, entao correm em sequencia com pouco overlap
    us01 = triangular(*historias["US-01 Alertas de Velocidade"])
    us03 = triangular(*historias["US-03 Score de Comportamento"])
    trilha_sw = (us01 + us03) / 1.3

    # Trilha hardware (comeca na semana 9)
    us04 = triangular(*historias["US-04 Sensor de Abertura de Bau"])
    us09 = triangular(*historias["US-09 Carga Refrigerada"])
    trilha_hw = HARDWARE_SEMANA + (us04 + us09) / 1.5

    # Projeto termina quando as duas trilhas terminam
    totais.append(max(trilha_sw, trilha_hw))

totais.sort()

def percentil(pct):
    return totais[int(pct * N / 100)]

media = sum(totais) / N

print("=" * 50)
print(f"Monte Carlo RouteWise — {N:,} simulacoes")
print("=" * 50)
print(f"P50 (cenario provavel):        {percentil(50):.1f} semanas")
print(f"P85 (compromisso com cliente): {percentil(85):.1f} semanas")
print(f"P95 (conservador):             {percentil(95):.1f} semanas")
print(f"Media:                         {media:.1f} semanas")
print("=" * 50)

File: modulo07-ferramentas-de-ia-para-gestao-de-projetos/modulo-04-estimativas-e-previsoes/monte-carlo-routewise.js
/*
 * Monte Carlo — RouteWise MVP
 * Modulo 4.2: Estimativas e Previsoes Probabilisticas
 *
 * Como rodar:
 * - Browser: abra o Console do DevTools, cole este arquivo e pressione Enter.
 * - Node.js: node monte-carlo-routewise.js
 */

const historias = {
  "US-01 Alertas de Velocidade": { o: 2, m: 3, p: 5 },
  "US-03 Score de Comportamento": { o: 3, m: 4, p: 6 },
  "US-04 Sensor de Abertura de Bau": { o: 1, m: 2, p: 3 },
  "US-09 Carga Refrigerada": { o: 2, m: 3, p: 5 },
};

const HARDWARE_SEMANA = 9; // US-04 e US-09 so comecam quando o hardware chega.
const N = 10_000; // Numero de simulacoes.

function triangular(o, m, p) {
  const u = Math.random();
  const fc = (m - o) / (p - o);

  if (u < fc) {
    return o + Math.sqrt(u * (p - o) * (m - o));
  }

  return p - Math.sqrt((1 - u) * (p - o) * (p - m));
}

const totais = [];

for (let i = 0; i < N; i += 1) {
  // Trilha software: comeca na semana 1.
  // US-03 depende de US-01, entao as duas correm em sequencia com pouco overlap.
  const us01 = triangular(
    historias["US-01 Alertas de Velocidade"].o,
    historias["US-01 Alertas de Velocidade"].m,
    historias["US-01 Alertas de Velocidade"].p
  );
  const us03 = triangular(
    historias["US-03 Score de Comportamento"].o,
    historias["US-03 Score de Comportamento"].m,
    historias["US-03 Score de Comportamento"].p
  );
  const trilhaSoftware = (us01 + us03) / 1.3;

  // Trilha hardware: so comeca na semana 9, quando o hardware chega.
  const us04 = triangular(
    historias["US-04 Sensor de Abertura de Bau"].o,
    historias["US-04 Sensor de Abertura de Bau"].m,
    historias["US-04 Sensor de Abertura de Bau"].p
  );
  const us09 = triangular(
    historias["US-09 Carga Refrigerada"].o,
    historias["US-09 Carga Refrigerada"].m,
    historias["US-09 Carga Refrigerada"].p
  );
  const trilhaHardware = HARDWARE_SEMANA + (us04 + us09) / 1.5;

  // O projeto termina quando a trilha mais lenta termina.
  totais.push(Math.max(trilhaSoftware, trilhaHardware));
}

totais.sort((a, b) => a - b);

function percentil(pct) {
  return totais[Math.floor((pct * N) / 100)];
}

const media = totais.reduce((soma, valor) => soma + valor, 0) / N;

console.log("=".repeat(50));
console.log(`Monte Carlo RouteWise — ${N.toLocaleString("pt-BR")} simulacoes`);
console.log("=".repeat(50));
console.log(`P50 (cenario provavel):        ${percentil(50).toFixed(1)} semanas`);
console.log(`P85 (compromisso com cliente): ${percentil(85).toFixed(1)} semanas`);
console.log(`P95 (conservador):             ${percentil(95).toFixed(1)} semanas`);
console.log(`Media:                         ${media.toFixed(1)} semanas`);
console.log("=".repeat(50));
